pkg/ingester/circuitbreaker_test.go

```diff
@@ -0,0 +1,816 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingester
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/failsafe-go/failsafe-go/circuitbreaker"
+	"github.com/go-kit/log"
+	"github.com/gogo/status"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/test"
+	"github.com/grafana/dskit/user"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/codes"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+func TestIsFailure(t *testing.T) {
+	t.Run("no error", func(t *testing.T) {
+		require.False(t, isFailure(nil))
+	})
+
+	t.Run("context cancelled", func(t *testing.T) {
+		require.False(t, isFailure(context.Canceled))
+		require.False(t, isFailure(fmt.Errorf("%w", context.Canceled)))
+	})
+
+	t.Run("gRPC context cancelled", func(t *testing.T) {
+		err := status.Error(codes.Canceled, "cancelled!")
+		require.False(t, isFailure(err))
+		require.False(t, isFailure(fmt.Errorf("%w", err)))
+	})
+
+	t.Run("gRPC deadline exceeded", func(t *testing.T) {
+		err := status.Error(codes.DeadlineExceeded, "broken!")
+		require.True(t, isFailure(err))
+		require.True(t, isFailure(fmt.Errorf("%w", err)))
+	})
+
+	t.Run("gRPC unavailable with INSTANCE_LIMIT details", func(t *testing.T) {
+		err := newInstanceLimitReachedError("broken")
+		require.True(t, isFailure(err))
+		require.True(t, isFailure(fmt.Errorf("%w", err)))
+	})
+
+	t.Run("gRPC unavailable with SERVICE_UNAVAILABLE details is not a failure", func(t *testing.T) {
+		stat := status.New(codes.Unavailable, "broken!")
+		stat, err := stat.WithDetails(&mimirpb.ErrorDetails{Cause: mimirpb.SERVICE_UNAVAILABLE})
+		require.NoError(t, err)
+		err = stat.Err()
+		require.False(t, isFailure(err))
+		require.False(t, isFailure(fmt.Errorf("%w", err)))
+	})
+
+	t.Run("gRPC unavailable without details is not a failure", func(t *testing.T) {
+		err := status.Error(codes.Unavailable, "broken!")
+		require.False(t, isFailure(err))
+		require.False(t, isFailure(fmt.Errorf("%w", err)))
+	})
+}
+
+func TestCircuitBreaker_IsActive(t *testing.T) {
+	var cb *circuitBreaker
+
+	require.False(t, cb.isActive())
+
+	registry := prometheus.NewRegistry()
+	cfg := CircuitBreakerConfig{Enabled: true, InitialDelay: 10 * time.Millisecond}
+	cb = newCircuitBreaker(cfg, false, log.NewNopLogger(), registry)
+	time.AfterFunc(cfg.InitialDelay, func() {
+		cb.setActive()
+	})
+	// When InitialDelay is set, circuit breaker is not immediately active.
+	require.False(t, cb.isActive())
+
+	// After InitialDelay passed, circuit breaker becomes active.
+	time.Sleep(2 * cfg.InitialDelay)
```

Comment:
This can lead to flaky tests on slow machines. We can use `require.Eventually` or `test.Poll` to wait for condition and give it more time.